### ISSUE TITLE
The generated certificate from dotnet dev-certs https isn't valid on iOS

### ISSUE THREAD

**Issue by lewcianci** [aspnet/AspNetCore#3957]
When you use dotnet dev-certs https --trust on Mac OSX, it creates a new certificate and adds it to the store as expected.  However, when you are using the iOS simulator to test that website, or even use Azure B2C authorisation, you can't use the generated certificate on the device.

The reason for this is the certificate generated isn't a self-signed cert with Basic Constraint x509 extension with enabled CA option.  The dotnet certs tool should provide the option to create this type of certificate.  Without it, you can't trust the certificate on iOS and you can't do any HTTPS calls from your iOS simulator to your localhost webserver.

**Comment by Tratcher** [comment]
@javiercn @blowdart

**Comment by lewcianci** [comment]
I also tried cloning the tool that makes the certificate to see if I could manually add the properties required but I couldn't make it work unfortunately :(.  This means that any https calls to a development server while developing won't work, because the iOS simulator does not share a certificate store with the parent OS (OSX).

**Comment by blowdart** [comment]
Can you not use a development mobile profile to trust it, using the Apple Configurator app? Changing the tools to start supporting the pretense of a CA is something I'm uncomfortable with to be honest. You wouldn't install a CA cert on a web site, so you're not testing with like for like type certs.

**Comment by lewcianci** [comment]
I'll try that and report back 😀

**Comment by javiercn** [comment]
Closing this issue as the question has been answered and there's no more action to be taken here.

**Comment by ckpearson** [comment]
@lewcianci did you get it working with the iOS simulator?

I'm running up against this issue, I have:

1. Created a profile in configurator
2. Added the certificate created by dev-certs
3. Signed the profile using my Apple Developer Credentials

iOS is happy to accept the profile, even going so far as giving it a checkmark to say it's a verifiably signed profile.

*BUT*

The trust store *still* won't let me trust the certificate (it simply doesn't show up).

Installing the profile:

<img width="343" alt="screenshot 2019-01-29 at 22 49 58" src="https://user-images.githubusercontent.com/5690115/51946068-65f6d100-2418-11e9-9386-8db1908285d0.png">

Certificates in profile:

<img width="341" alt="screenshot 2019-01-29 at 22 51 53" src="https://user-images.githubusercontent.com/5690115/51946171-ae15f380-2418-11e9-80a2-18ecefdf4304.png">

Profile has been installed:

<img width="344" alt="screenshot 2019-01-29 at 22 54 28" src="https://user-images.githubusercontent.com/5690115/51946240-def62880-2418-11e9-9928-cf3375686ebb.png">

The certificate trust store is empty however:

<img width="345" alt="screenshot 2019-01-29 at 22 56 58" src="https://user-images.githubusercontent.com/5690115/51946387-3399a380-2419-11e9-8422-7edbad081739.png">

This is a bit of an oversight in my opinion, as understandably iOS wants to be able to maintain a trust chain for the certificate, and it can't do that because there's no CA information.

The only potential workaround I've been able to identify (and yet to pursue) would be to generate my own CA, produce certificates using that, get .net to use those certs, and then send the CA & dev cert over to iOS - but this seems like overkill considering dev-certs is *supposed* to support certificate-related tasks for development.

I'd request you reopen the issue and look into allowing the tool to generate suitable certificates for these sorts of scenarios.

**Comment by lewcianci** [comment]
No I never got it working. I agree with you @ckpearson - these development scenarios are supposed to make it super easy for development in these situations.  I never found a way forward.  I think the only reason why I had these issues is because I was trying to use oauth authentication in my app and I absolutely had to have https on the identity server.

I'd love to see this fixed, or worked around, or anything

**Comment by lewcianci** [comment]
@blowdart any chance of this getting some attention?  @ckpearson  has included screenshots and a fresh set of reasons as to why this is required.  If I had the slightest idea how the dev-certs app worked I would have a go at it and submit a PR.

**Comment by ckpearson** [comment]
@lewcianci I'm building something using IdentityServer, so I of course want https.

It's all fine for the web app, because that's running on my local machine and as such is able to just trust the certificate because it's in the keychain.

iOS is a bit stricter about it because it can't verify the CA, we either need dev-certs to produce a local-dev CA cert **and** the local-dev signing cert, or just produce a cert with the CA extension which is what iOS is looking for / a valid chain to enable you to trust it.

I gather that the common dev scenario for .net core is web apps, but for native app dev in simulators and on test devices, this really is needed...

**Comment by lewcianci** [comment]
FWIW, I think its fairly straightforward to get ASP .NET to load a HTTPS certificate at startup that could be provisioned from your development CA.  I started going down this path back in the day but got waylaid by other stuff.

I'd love to see this get fixed 👍

**Comment by ckpearson** [comment]
@lewcianci tbh who has the time?! I had a bloody hard enough time getting my head around OAuth and OIDC for IdentityServer, and then the IdentServer docs are hardly up to scratch, they're out of date in places, and there's no real section on how to configure clients for native app scenarios 🤷‍♂️ 

Still, if it were easy, we wouldn't have such in-demand jobs.

**Comment by ckpearson** [comment]
@lewcianci can confirm:

<img width="340" alt="screenshot 2019-01-29 at 23 54 50" src="https://user-images.githubusercontent.com/5690115/51948681-5cbe3200-2421-11e9-949c-490476eb4013.png">

<img width="350" alt="screenshot 2019-01-29 at 23 55 10" src="https://user-images.githubusercontent.com/5690115/51948691-647dd680-2421-11e9-9f5e-8165f775f510.png">

The difference here is that I've set `CA:true` manually by generating my own cert, but I can see in the source for dev-certs: https://github.com/aspnet/Extensions/blob/8ddc5722596937430ddb6b2aec2fcea06723d5a7/shared/Microsoft.AspNetCore.Certificates.Generation.Sources/CertificateManager.cs#L169 that it would be possible for it to do the same...

**Comment by ckpearson** [comment]
Actually, the CA thing just creates a root cert, which (I think) is useless in and of itself, so I'm now trying a workaround to produce a CA root and derivative certs myself, will update here as could be a way around this for people before a fix is made / it doesn't get changed.

**Comment by lewcianci** [comment]
Yeah thats exactly it @ckpearson (as an aside, I agonised for days over OIDC and OpenIddict.  I got it working in the end but I felt extremely dense the whole time (like everyone else has it working/gets it working, why can't I?))

**Comment by ckpearson** [comment]
@lewcianci I can confirm I now have this up and running!

<img width="339" alt="screenshot 2019-01-30 at 00 52 46" src="https://user-images.githubusercontent.com/5690115/51950687-66e42e80-2429-11e9-94b3-f43b9ea5189c.png">

I've put the instructions I followed [here](https://gist.github.com/ckpearson/f96074697c583de0d2599c6009191824)

It's longer-winded than I'd have liked, and it sucks that I have to configure the cert at runtime, but it does the trick!

**Comment by blowdart** [comment]
A work around that uses a self trusted root CA isn't something I would be happy putting in the dev tool, because, well, we're venturing into risky territory, as now you have a root CA that can be used for more than local host. You can generate such a set in powershell, but it's not something we'd do.

**Comment by ckpearson** [comment]
@blowdart really? Deploying **every** time you want to test? Isn't that a bit mental? And you can't use certs on Azure unless it's a paid tier (last I checked).

I honestly feel it's an oversight for this to not be supported, at the very least with huge flashy warnings and all manner of controls around it, other tools like Fiddler support these sorts of certificates for inspecting traffic from an IOS device.

What specifically would be the concern around generating a new root CA? Particularly one limited to an individual dev machine, for work on localhost, with a password the user has to enter in the console?

I just don't see how (and this could be my lack of knowledge, if so, please do let me know) this could be abused, at least, not in any far-reaching, super-scary way...

**Comment by ckpearson** [comment]
@blowdart and I'd add that the OIDC server I'm developing against is IdentityServer, running locally...

**Comment by lewcianci** [comment]
Your solution to push to the cloud is a nonstarter because:

- I would then have to pay for cloud hosting costs
- It would complicate my development scenario to have to push over and over to the cloud to test these things (not to mention things like database migrations etc)

I didn't realise that developing with dotnet would leave me with with broken development workflows due to how heavily opinionated the tooling is (we're never going to let you generate CA's, even though you need them for a legitimate development scenario, because we think its a bad idea)

@ckpearson  the concern is that your CA could get out in the wild and that would be bad for security.  Something like, someone would deploy your code into production, see the cert warnings, and just install the certs from the development environment into production to make the errors "go away".  That's really bad for a lot of reasons.  But it depends on many stuff-ups of many different people first.

**Comment by ckpearson** [comment]
@lewcianci oh, I'd argue no more so than secrets hardcoded in OSS repos anyway 🤷‍♂️

**Comment by blowdart** [comment]
Against every other piece of security work we have this is very low in the priority list, as well as being dangerous. As there are other ways to create suitable certs I'm just not convinced about the safety or that I should cancel something else to do this instead. Which is what it would boil down to.

**Comment by ckpearson** [comment]
@blowdart fair enough, is there somewhere I can log this as something for further consideration down the line / something that ought to at the least be documented, as I ran headfirst into this working with IdentityServer and it having practically zero documentation around mobile app dev scenarios.

I suppose I'm trying to work out what repo / team would be responsible for at least looking at this, as I feel it's something the community could do with being aware of / having input on.

**Comment by lewcianci** [comment]
@blowdart  that's okay, as long as its somewhere in your priority list.  Also you keep saying its dangerous, could you elaborate?  @ckpearson  and I have highlighted a legitimate development scenario for this, if you flesh out why its a bad idea I'll take heed appropriately.  As far as I know, you are concerned that a CA could get out in the wild and that could cause security issues?

Also, you don't have to cancel anything else, it can just go on the backburner.  @ckpearson  or myself could submit a PR to bring this functionality in.

**Comment by blowdart** [comment]
I'll reopen in issue and put it into the backlog for consideration in the 3.1 timeframe.

**Comment by PinpointTownes** [comment]
> (as an aside, I agonised for days over OIDC and OpenIddict. I got it working in the end but I felt extremely dense the whole time (like everyone else has it working/gets it working, why can't I?))

If you want to chat about your (negative) experience with OpenIddict and/or have concrete elements you think could/should be improved (I suspect documentation is one of them), feel free to ping me on Gitter (https://gitter.im/openiddict/openiddict-core) or on Slack (http://tattoocoder.com/asp-net-core-slack-community/)

**Comment by blowdart** [comment]
And for triage;

Any root CA generated would need to be very constrained and only able to issue HTTPS certificates via EKU, not wide open for any purpose, so EKU of 1.3.6.1.5.5.7.3.1. Furthermore there should be a path length constraint of 0. Key usage should be restricted to Certificate Signing. As there will be no CRL or OCSP there's no need for those uses. 

Validity length to be discussed.

Would this feature be offset by lets encrypt support instead? (Probably not)

**Comment by ckpearson** [comment]
@blowdart I understood practically none of that, but what might be a solution would be to use lets encrypt certs as you said, I'm fairly sure iOS would trust those quite readily, and avoids the need for self-issuance and self-creation of root CAs.

**Comment by blowdart** [comment]
You'd never get a lets encrypt certificate for localhost. So you'd have to get a domain validated cert, and the normal way is by putting a file on the server you want a cert for, but because that's your dev PC that's going to be firewalled off, so now you have to go the DNS route to validate your request, which can't be automated, because all DNS providers have different APIs. Hence "probably not", and why let's encrypt isn't really a solution for any dev work.

**Comment by ckpearson** [comment]
@blowdart gotcha, so perhaps the best approach really is to abandon local-dev https and just ensure it's configured and utilised properly upon deployment, because there's still all the considerations to consider like running on a live test device over the local network too, as the cert isn't going to match "localhost" when requested remotely anyway...

'tis a shame that secure local dev practices aren't easier than this.

**Comment by blowdart** [comment]
And we had that before, but by enabled HTTPS for the majority of scenarios we're helping to close a gab and not rely on devs remembering to do the right configuration, which is a win. We won't abandon local-dev https simply because it won't work on iOS.

**Comment by ckpearson** [comment]
@blowdart apologies, I meant I *could* abandon local https for my dev scenario, rather than dropping it as a default in the templates - it most certainly should be default in the templates!

**Comment by 0xced** [comment]
I'm currently working on a solution for iOS simulators to trust the ASP.NET Core development certificate. I got it to work but it needs cleanup, better diagnostics and error handling before I submit a pull request. I have not yet investigated the possibility of trusting the development certificate on an actual iOS device plugged in USB.

You can see my early work in progress on https://github.com/0xced/AspNetCore/tree/dev-certs-https-ios

**Comment by 0xced** [comment]
Also, if you want to go the CA route you should have a look at [mkcert](https://github.com/FiloSottile/mkcert) which is *a simple zero-config tool to make locally trusted development certificates with any names you'd like.*

**Comment by rmarinho** [comment]
Plus one on maybe try to make the tool work for mobile and iOS. We want to make developers a delightful experience developing cross platform solutions with .NET.
Since https as default is the way to go, would be great if we can make it work with a flag when generating the dev cert if we want to make it work for mobile too. 
Other approach is to have documentation how to get this working for iOS and Android .

**Comment by 0xced** [comment]
@rmarinho I'll have to rebase #14228 on master then get it merged before I can continue to work on this feature. But first I'd like #19483 to be merged in order to reduce conflicts when rebasing and also to be sure that there's traction for this feature.

**Comment by DTX-Elliot** [comment]
Any news on this? Has anyone managed to get their ios simulator to connect to localhost?
